### ISSUE TITLE
"test if bootstrap_wait_for_host is set correctly" fails

### ISSUE THREAD

**Issue by quotengrote** [robertdebock/ansible-role-bootstrap#43]
## Description
The run fails because a conditionals check can't executed.
The bug was introduced in [005c7e0](https://github.com/robertdebock/ansible-role-bootstrap/commit/005c7e0ed4fc1dac9fef879e5b15fd388083797e)

```yaml
---
  - hosts: testeinzeln
    become: yes
    gather_facts: false
    max_fail_percentage: 20%

    roles:
      - { role: robertdebock.bootstrap, tags: "bootstrap" }
      - { role: ryandaniels.create_users, tags: "user", become: yes }
      - { role: nickjj.ansible-user, tag: "ansible", become: yes }

    vars:
      ### nickjj.ansible-users
      user_name: "ansible-user"
      user_local_ssh_key_path: "/home/mg/ansible/id_rsa_ansible_user_pub"
      user_enable_passwordless_sudo: True
      ### reobertdebock.bootstrap
      bootstrap_user: mg
      bootstrap_wait_for_host: no
      bootstrap_timeout: 1
      bootstrap_retries: 1
      ### ansible
      ansible_user: "mg"
      ansible_password: "XXXXXXXXXXXX"
      ansible_become_password: "XXXXXXXXXXXX"
      ansible_ssh_common_args: "'-o StrictHostKeyChecking=no'"

# Nach dem ersten durchlaufen ist keine Anmeldung mehr per Passwort & ssh möglich. Somit scheitert auch der Versuch das Playbook ein zweites mal durchlaufen zu lassen.

```
--> [Complete Repo](https://github.com/quotengrote/ansible)

## Output
```bash
mg@ansible:~/ansible$ ansible-playbook playbooks/base/1_bootstrap.yml -i inventories/inventory

PLAY [testeinzeln] **************************************************************************************************************************************************************************

TASK [robertdebock.bootstrap : include assert.yml] ******************************************************************************************************************************************
included: /home/mg/ansible/roles/robertdebock.bootstrap/tasks/assert.yml for vm-test.grote.lan

TASK [robertdebock.bootstrap : test if bootstrap_user is set correctly] *********************************************************************************************************************
ok: [vm-test.grote.lan]

TASK [robertdebock.bootstrap : test if bootstrap_wait_for_host is set correctly] ************************************************************************************************************
fatal: [vm-test.grote.lan]: FAILED! => {"msg": "The conditional check 'bootstrap_wait_for_host is boolean' failed. The error was: template error while templating string: no test named 'boolean'. String: {% if bootstrap_wait_for_host is boolean %} True {% else %} False {% endif %}"}

NO MORE HOSTS LEFT **************************************************************************************************************************************************************************

NO MORE HOSTS LEFT **************************************************************************************************************************************************************************

PLAY RECAP **********************************************************************************************************************************************************************************
vm-test.grote.lan          : ok=2    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

## Environment

- Control node OS: 
```
mg@ansible:~/ansible$ uname -a
Linux ansible 5.4.65-1-pve #1 SMP PVE 5.4.65-1 (Mon, 21 Sep 2020 15:40:22 +0200) x86_64 x86_64 x86_64 GNU/Linux
mg@ansible:~/ansible$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.1 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.1 LTS"
VERSION_ID="20.04"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
mg@ansible:~/ansible$ ansible --version
ansible 2.10.1
  config file = /home/mg/ansible/ansible.cfg
  configured module search path = ['/home/mg/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/mg/.local/lib/python3.8/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.8.5 (default, Jul 28 2020, 12:59:40) [GCC 9.3.0]

```
- Managed node OS: 
```
mg@lxc-test:~/.ssh$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.1 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.1 LTS"
VERSION_ID="20.04"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

**Comment by robertdebock** [comment]
Hi, @quotengrote. Thanks for your issue.

I'm intrigued by this issue; the [default](https://github.com/robertdebock/ansible-role-bootstrap/blob/master/defaults/main.yml) value is `no`. You could leave the variable out of your playbook, although it looks completely fine. You may experiment using `False` or `0` instead of `no`.

The default settings (and a few tasks to troubleshoot the value):

```
    TASK [ansible-role-bootstrap : show bootstrap_wait_for_host] *******************
    ok: [bootstrap-fedora-latest] => 
      msg: 'False'
    
    TASK [ansible-role-bootstrap : show type of bootstrap_wait_for_host] ***********
    ok: [bootstrap-fedora-latest] => 
      msg: bool
    
    TASK [ansible-role-bootstrap : test if bootstrap_wait_for_host is set correctly] ***
    ok: [bootstrap-fedora-latest]
```

The variable `bootstrap_retries: 1` is not in use anymore, you could also leave it out.

I see you are pinning to a commit-hash. I sometimes experiment, but when everything is working, I release a version. I would pin to a version rather than a commit-hash. I've just released 4.3.5.

Please let me know if changing to `False` over `no` or updating works for you, although I don't see how.

**Comment by quotengrote** [comment]
> The variable `bootstrap_retries: 1` is not in use anymore, you could also leave it out.

removed


> I see you are pinning to a commit-hash. I sometimes experiment, but when everything is working, I release a version. I would pin to a version rather than a commit-hash. I've just released 4.3.5.

For the test i pinned this version, but...

> Please let me know if changing to `False` over `no` or updating works for you, although I don't see how.

neither false, no or 0 did the trick

**Comment by robertdebock** [comment]
Thanks for the feedback, what a weird issue, can't really reproduce it.

I typically test values using [this types role](https://github.com/robertdebock/ansible-role-types/). In there, `no` [returns as a boolean](https://github.com/robertdebock/ansible-role-types/runs/1454174764?check_suite_focus=true#step:6:295).

I'm not sure how to try to reproduce your situation. Maybe try that `robertdebock.types` role, see how it reacts.

I've given it a try on my [ansible-tester](https://github.com/robertdebock/ansible-tester/blob/master/bootstrap.yml) repo, seems to work.

**Comment by quotengrote** [comment]
Hi, i've tried your [Types-Role](https://github.com/robertdebock/ansible-role-types/). 
Its the same error like the bootstrap role.
<s>I'll build my ansible-host new.</s> Same error on a completely new host.

## Log
```
mg@ansible:~/ansible$ ansible-playbook playbooks/on-off/types.yml -i inventory --limit acng.grote.lan

PLAY [all] *******************************************************************************************

TASK [Gathering Facts] *******************************************************************************
ok: [acng.grote.lan]

TASK [ansible-role-types : include assert.yml] *******************************************************
included: /home/mg/ansible/roles/ansible-role-types/tasks/assert.yml for acng.grote.lan

TASK [ansible-role-types : test if types_strings is set correctly] ***********************************
ok: [acng.grote.lan]

TASK [ansible-role-types : test if types_integers is set correctly] **********************************
ok: [acng.grote.lan]

TASK [ansible-role-types : test if types_booleans is set correctly] **********************************
ok: [acng.grote.lan]

TASK [ansible-role-types : test if types_floats is set correctly] ************************************
ok: [acng.grote.lan]

TASK [ansible-role-types : show strings] *************************************************************
ok: [acng.grote.lan] => (item=hello) => {
    "msg": "hello is type str"
}
ok: [acng.grote.lan] => (item=1.2.3) => {
    "msg": "1.2.3 is type str"
}

TASK [ansible-role-types : test strings] *************************************************************
ok: [acng.grote.lan] => (item=hello) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": "hello",
    "msg": "All assertions passed"
}
ok: [acng.grote.lan] => (item=1.2.3) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": "1.2.3",
    "msg": "All assertions passed"
}

TASK [ansible-role-types : show integers] ************************************************************
ok: [acng.grote.lan] => (item=0) => {
    "msg": "0 is type int"
}
ok: [acng.grote.lan] => (item=1) => {
    "msg": "1 is type int"
}
ok: [acng.grote.lan] => (item=2) => {
    "msg": "2 is type int"
}
TASK [ansible-role-types : test integers] ************************************************************
ok: [acng.grote.lan] => (item=0) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": 0,
    "msg": "All assertions passed"
}
ok: [acng.grote.lan] => (item=1) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": 1,
    "msg": "All assertions passed"
}
ok: [acng.grote.lan] => (item=2) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": 2,
    "msg": "All assertions passed"
}

TASK [ansible-role-types : show booleans] ************************************************************
ok: [acng.grote.lan] => (item=True) => {
    "msg": "True is type bool"
}
ok: [acng.grote.lan] => (item=True) => {
    "msg": "True is type bool"
}
ok: [acng.grote.lan] => (item=True) => {
    "msg": "True is type bool"
}
ok: [acng.grote.lan] => (item=True) => {
    "msg": "True is type bool"
}
ok: [acng.grote.lan] => (item=True) => {
    "msg": "True is type bool"
}
ok: [acng.grote.lan] => (item=True) => {
    "msg": "True is type bool"
}
ok: [acng.grote.lan] => (item=False) => {
    "msg": "False is type bool"
}
ok: [acng.grote.lan] => (item=False) => {
    "msg": "False is type bool"
}
ok: [acng.grote.lan] => (item=False) => {
    "msg": "False is type bool"
}
ok: [acng.grote.lan] => (item=False) => {
    "msg": "False is type bool"
}
ok: [acng.grote.lan] => (item=False) => {
    "msg": "False is type bool"
}
ok: [acng.grote.lan] => (item=False) => {
    "msg": "False is type bool"
}
TASK [ansible-role-types : test booleans] ************************************************************
fatal: [acng.grote.lan]: FAILED! => {"msg": "The conditional check 'item is boolean' failed. The error was: template error while templating string: no test named 'boolean'. String: {% if item is boolean %} True {% else %} False {% endif %}"}

PLAY RECAP *******************************************************************************************
acng.grote.lan             : ok=11   changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

```

**Comment by robertdebock** [comment]
Okay, I'm starting to get a feeling:

`The error was: template error while templating string: no test named 'boolean'`

This is an assumption: Your version of jinja does not have the test `boolean`. I'm guessing you're using an unexpected version of Jinja.

On my system `pip show Jinja2` shows:

```
Name: Jinja2
Version: 2.11.2
Summary: A very fast and expressive template engine.
Home-page: https://palletsprojects.com/p/jinja/
Author: Armin Ronacher
Author-email: armin.ronacher@active-4.com
License: BSD-3-Clause
Location: /home/robertdb/.local/lib/python3.9/site-packages
Requires: MarkupSafe
Required-by: molecule, jinja2-time, cookiecutter, click-completion, ansible-base
```

**Can you tell me what version of Jinja you have?**

In my installation I can find jinja2/test.py with a contents of: (partial, snippet)

```
TESTS = {
# STUFF REMOVED
    "boolean": test_boolean,
# STUFF REMOVED
```

**Comment by quotengrote** [comment]
Looks like it.
## Before
```
mg@ansible:~/ansible$ pip show Jinja2
Name: Jinja2
Version: 2.10.1
Summary: A small but fast and easy to use stand-alone template engine written in pure python.
Home-page: http://jinja.pocoo.org/
Author: Armin Ronacher
Author-email: armin.ronacher@active-4.com
License: BSD
Location: /usr/lib/python3/dist-packages
Requires:
Required-by: ansible-base

```
## After Update to...
```
mg@ansible:~/ansible$ pip show Jinja2
Name: Jinja2
Version: 2.11.2
Summary: A very fast and expressive template engine.
Home-page: https://palletsprojects.com/p/jinja/
Author: Armin Ronacher
Author-email: armin.ronacher@active-4.com
License: BSD-3-Clause
Location: /home/mg/.local/lib/python3.8/site-packages
Requires: MarkupSafe
Required-by: ansible-base
```

### types
```
TASK [ansible-role-types : show booleans] ***************************************************************************************************************************************************
ok: [acng.grote.lan] => (item=True) => {
    "msg": "True is type bool"
}
ok: [acng.grote.lan] => (item=True) => {
    "msg": "True is type bool"
}
ok: [acng.grote.lan] => (item=True) => {
    "msg": "True is type bool"
}
ok: [acng.grote.lan] => (item=True) => {
    "msg": "True is type bool"
}
ok: [acng.grote.lan] => (item=True) => {
    "msg": "True is type bool"
}
ok: [acng.grote.lan] => (item=True) => {
    "msg": "True is type bool"
}
ok: [acng.grote.lan] => (item=False) => {
    "msg": "False is type bool"
}
ok: [acng.grote.lan] => (item=False) => {
    "msg": "False is type bool"
}
ok: [acng.grote.lan] => (item=False) => {
    "msg": "False is type bool"
}
ok: [acng.grote.lan] => (item=False) => {
    "msg": "False is type bool"
}
ok: [acng.grote.lan] => (item=False) => {
    "msg": "False is type bool"
}
ok: [acng.grote.lan] => (item=False) => {
    "msg": "False is type bool"
}

TASK [ansible-role-types : test booleans] ***************************************************************************************************************************************************
ok: [acng.grote.lan] => (item=True) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": true,
    "msg": "All assertions passed"
}
ok: [acng.grote.lan] => (item=True) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": true,
    "msg": "All assertions passed"
}
ok: [acng.grote.lan] => (item=True) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": true,
    "msg": "All assertions passed"
}
ok: [acng.grote.lan] => (item=True) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": true,
    "msg": "All assertions passed"
}
ok: [acng.grote.lan] => (item=True) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": true,
    "msg": "All assertions passed"
}
ok: [acng.grote.lan] => (item=True) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": true,
    "msg": "All assertions passed"
}
ok: [acng.grote.lan] => (item=False) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": false,
    "msg": "All assertions passed"
}
ok: [acng.grote.lan] => (item=False) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": false,
    "msg": "All assertions passed"
}
ok: [acng.grote.lan] => (item=False) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": false,
    "msg": "All assertions passed"
}
ok: [acng.grote.lan] => (item=False) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": false,
    "msg": "All assertions passed"
}
ok: [acng.grote.lan] => (item=False) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": false,
    "msg": "All assertions passed"
}
ok: [acng.grote.lan] => (item=False) => {
    "ansible_loop_var": "item",
    "changed": false,
    "item": false,
    "msg": "All assertions passed"
}


```

tl;dr

It works now with **Jinja2 - Version: 2.11.2** and **ansible 2.10.1**. THX

**Comment by robertdebock** [comment]
Wow, I've never seen that issue before. That's worth a mention in the "requirements" section of the README.md.

**Comment by DavidDessertine** [comment]
Hi,
I encounter the same issue, on AWX running playbooks on RHEL7 and RHEL8.
It's the latest major version of AWX (15.0) and Jinja is in version 2.11.2 (after manuel upgrade) and ansible in v 2.9.15 on Centos 8.3.2011, can't install 2.10.
Can you see if a fix is possible ?
Thanks

**Comment by robertdebock** [comment]
Hi @Climberdav, I will update the documentation and introduce a `requirements.txt` file that describe these requirements.

In case you can't update the controller, you can always refer to an older version of the role, until the controller has been updated.

Sorry that I'm unable to help you directly, hope the older version help you out.
